storage/innobase/srv/srv0start.cc

```diff
@@ -1988,6 +1985,22 @@ void innodb_preshutdown()
 
   if (srv_read_only_mode)
     return;
+
+#ifndef EMBEDDED_LIBRARY
+  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && srv_was_started &&
+      !trx_rollback_is_active && trx_sys.any_active_transaction_recovered())
+  {
+    /* Trx rollback thread may not be started yet, since error happened before
+       innobase_tc_log_recovery_done(), thus it is started here to finish the
+       normal shutdown process.
+     */
+    ib::warn() << "Start rollback task at preshutdown to rollback active "
+                  "transactions recovered.";
+    trx_rollback_is_active= true;
+    srv_thread_pool->submit_task(&rollback_all_recovered_task);
+  }
+#endif
+
   if (!srv_fast_shutdown && srv_operation <= SRV_OPERATION_EXPORT_RESTORED)
     if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && srv_was_started)
       while (trx_sys.any_active_transactions())
```

Comment:
Should we invoke the above only if the below condition (shutdown with `innodb_fast_shutdown=0`) holds? In that case, would `sql_print_information()` messages be visible?

@vaintroub Is it intentional that `sql_print_information()` messages would be suppressed during a normal shutdown? I think that the message about starting rollback (if it is needed at all) should be informational and not a warning, but apparently the message would be suppressed during a normal shutdown.

Reply:
it is disabled by `disable_log_notes= 1;` in function unireg_abort() which is a server layer code, it is called only if error happens at startup.

Reply:
That turns out to be an intentional change that was made in affff1aefc81ea0a3cb778a6617504ccd1e905fc. Maybe it was not realized back then that `unireg_abort()` is being invoked as part of a normal shutdown,with no prior errors. In such a case, it would seem to be helpful to not filter out any `sql_print_information()` messages.

In any case, we may ‘sidestep’ this by not adding a message at all, and by starting the `rollback_all_or_recovered_task` right above the `while` loop in the following code snippet.

Reply:
removed the warning.

Reply:
> Should we invoke the above only if the below condition (shutdown with innodb_fast_shutdown=0) holds?  we have to start the rollback task if innodb_fast_shutdown is enabled. because logs_empty_and_mark_files_at_shutdown() will wait even innodb_fast_shutdown is enabled. It depends on rollback thread to deregister all active transactions in case innodb_fast_shutdown is enabled.